### DISCUSSION TITLE
[2.0.13] Autorestart of Workflowinstances for multiple Workflowrunners

### DISCUSSION THREAD

**Discussion by dakuchar** [elsa-workflows/elsa-core#4546] [Q&A]
Hi,

I have a question regarding the autoresume of workflows in the state "Running" on Startup. I would like to know wether you consider the issues I describe below as an actual issue with Elsa or if we're using Elsa not the way it was intended and our setup is the problem.

So here's our setup:
- we have multiple different Workflowrunners, that each use Elsa Workflow Engine
- these workflowrunners are redundant, there are 2 instances for each one
- we use SQL Server for persistence. The redundant workflow runners always share the same Database. Some different Workflowrunners also share the same DB
- A few of these workflows are long running, so we cancel those on Shutdown via CancellationToken
 
Now let me describe the issues, that we get with this setup:

First of all: Cancellation works fine. The result is that the cancelled Workflows are left in the Workflow Status "Running" in the database and can be continued from the last persisted state, which is good.

When any workflowrunner is restarted Elsa will now try to continue all Workflows it finds in the database in the Status Running from the last persisted state. This works fine in the Single Instance Usecase, but when having multiple Instances of the same Workflowrunner / different Workflowsrunners it causes issues:

1) It will try to resume workflows currently running on another workflowrunner instance
2) It will try to resume workflows from a completely different workflowrunner, if these are stored in the same database (minor Problem as this only causes Errors in the Logfile, since the Resumption just fails)
3) If multiple instances of the workflow runner are started at the same time, they will all resume the same workflows

I checked, if using DistributedLocking solves the issue, but DistributedLocking only delays Workflow Resumption. It will keep the state of the Workflow in Memory and continue after the distributed lock has been released. Also in this case it will not refresh the workflow state and just continue with the state, that the workflow had, when it checked the DistributedLock. When it is now in the State "Finished" or "Failed" workflow resumption still happens.

**Comment by sfmskywalker** [comment]
Hi @dakuchar ,

The symptoms you describe do sound like an issue. The intent is to acquire a distributed lock when an application instance starts, and then resume all workflows that are in the Running state. The lock is there to prevent a race condition where multiple containers start at around the same time, where each subsequent container should simply not try and resume workflow instances if they fail to acquire the lock.

From your description, it sounds like that there are two issues: one expected due to a design flaw, and one unexpected:

1. Unexpected: when multiple application instances start, they should all attempt to acquire a distributed lock, and only one of them should succeed, allowing it to resume Running workflow instances. From your description, it sounds like this isn't what's happening. One reason might be that the distributed lock provider isn't configured to use a shared resource. Out of the box, it uses a local file - which would mean that each application instance will attempt to acquire a lock on its own file system - which will always succeed.
2. Expected due to design flaw: new container instances that attempt to resume workflows with the Running status. The problem you observed is when another instance acquired a lock on this workflow instance in order to execute it - which will put its status into Running. However, the new app instance attempting to resume this workflow should fail to do so, since there should be a lock on that workflow instance.

I wasn't aware of the design flaw until you described it in this issue, and is something I'll give some more thought for Elsa 3. Once we have a good design, we might be able to back-port it to Elsa 2. And, of course, we're open to any suggestions you may have in the meantime.

**Reply by dakuchar** [reply]
Hi @sfmskywalker,

thanks for taking a look. From my tests the main issue with the distributed lock using SqlServer was that it didn't actually fail to acquire a lock. It just waited until the other process released it and then continued with the resumption at the state the workflow was in when it first tried to resume it.

I took a quick glance at the source code and the distributed lock behavior does seem to be configurable, but I don't currently see the configuration exposed on the option builder, when adding a distributed lock.
Also I think the retry behavior is generally a good thing when trying to acquire a lock, because it resolves problems like the network being not available for a few seconds and adds some redundancy, if an engine tries to resume a workflow and somehow fails to do so, because then another engine can try.

My test scenario was the following:

- I had 3 workflows in the DB, that were left in the Status 'Running'
- I added DistributedLocks with UseSqlServer in ElsaOptionsBuilder
- I ran an application that ran Elsa's autostart twice in parallel for simulating having multiple applications start at the same time
- I set up the resumed activity to complete with a 'Done' Outcome after 5 seconds in each resumed Workflow

This resulted in the following behavior: 

- All 3 Workflows were resumed on startup.
- As soon as the 3 activites were 'Done' and the Workflows finished, the 3 Workflows were resumed again and the activities were executed again, even though the Workflow and the activities were finished now.


In order to keep the retry behavior as it currently is, I would suggest the following change to Elsa's current behavior:
After the lock has been acquired on workflow resumption, the engine reloads the Workflow Instance to check, if the state was altered by another worklflow runner.
If it's still 'Running' it will resume from its current state, otherwise it will not resume.

**Reply by sfmskywalker** [reply]
I totally agree with your suggestion - after a workflow instance has executed inside of a lock, it is bound to have changed, which means that before the next process or thread starts executing it after getting their turn, they first need to load a fresh copy from the database.

If you submit a PR I can merge it relatively soon if you want.

**Reply by dakuchar** [reply]
Yes, I'll take a look into making a PR this week. Thanks for the discussion
